src/main/java/picard/sam/SplitSamBySize.java

```diff
@@ -101,29 +106,52 @@ protected int doWork() {
-        final int readsPerFile = (int) Math.ceil(TOTAL_READS_IN_INPUT / (double) splitToNFiles);
-        final SAMFileWriter[] writers = generateWriters(factory, header, splitToNFiles);
+        final long totalCompressedBytes = INPUT.length();
+        final int inputBytesPerFile = (int) Math.ceil(totalCompressedBytes / (double) SPLIT_TO_N_FILES);
+        final SAMFileWriter[] writers = generateWriters(factory, header, SPLIT_TO_N_FILES);
 
-        int readsWritten = 0;
         int writerIndex = 0;
         String lastReadName = "";
+        ArrayList<SAMRecord> blockRecords = new ArrayList();
+        long bufferedBlockEnd = 0;
+        long bufferedBlockStart = 0;
+        final SAMRecordIterator iterator = reader.iterator();
         final ProgressLogger progress = new ProgressLogger(log);
-        for (final SAMRecord currentRecord : reader) {
-            if (readsWritten >= readsPerFile && !lastReadName.equals(currentRecord.getReadName())) {
-                writerIndex++;
-                readsWritten = 0;
+        // Iterate through the input file, buffer each block's worth of reads so we can split within a block
+        while (iterator.hasNext()) {
+            final SAMRecord currentRecord = iterator.next();
+            final long currentBlockEnd;
+            try {
+                currentBlockEnd = inStream.position();
+            } catch (IOException e) {
+                e.printStackTrace();
+                return 1;
+            }
+            if (!iterator.hasNext()) {
+                // This is the last record of the bam, add it to the previous block so it gets written out
+                blockRecords.add(currentRecord);
             }
-            writers[writerIndex].addAlignment(currentRecord);
-            lastReadName = currentRecord.getReadName();
-            readsWritten++;
-            progress.record(currentRecord);
-        }
 
-        if (progress.getCount() != TOTAL_READS_IN_INPUT) {
-            log.warn(String.format("The TOTAL_READS_IN_INPUT (%s) provided does not match the reads found in the " +
-                            "input file (%s).", TOTAL_READS_IN_INPUT, progress.getCount())
-                   );
+            // We finished the block, write it out
+            if (currentBlockEnd != bufferedBlockEnd || !iterator.hasNext()) {
+                int idxInBlock = 0;
+                final long bufferedBlockSize = bufferedBlockEnd - bufferedBlockStart;
+                for (final SAMRecord record : blockRecords) {
+                    // Estimate the current position in bytes by subdividing the buffered block by the number of records in it
+                    final long currentPosition = bufferedBlockStart + ((idxInBlock * bufferedBlockSize) / blockRecords.size());
+                    if (currentPosition >= inputBytesPerFile * (writerIndex + 1) && !lastReadName.equals(record.getReadName())) {
+                        writerIndex++;
```

Comment:
if your estimation is wrong, would you get more than N_FILES files? or possibly less? it would be good to protect against this somehow (not change the writer if it's already N_FILES-1 and also emit an warning if the actual number of files with data differs from the requested number. )

Reply:
I have to say that I'm having a tough time following the logic used here. can you explain how it works, in a comment perhaps..